### PR TITLE
Add 'exclude_addons' param to search API

### LINKED ISSUE #6376: Add exclusion field on Search API
Hi, the addons-frontend has to display a set of related add-ons made by any of the authors on a add-on detail page (see: https://github.com/mozilla/addons-frontend/pull/3027 and https://github.com/mozilla/addons-frontend/pull/3083). It would be unfortunate to have the add-on of the detail page also listed in the "other add-ons by these authors". I do the filtering on the client side right now, but it will not play nice with pagination and I think we will need such filters sooner or later.

Would it be possible to add exclusion filters to the [Search API](https://addons-server.readthedocs.io/en/latest/topics/api/addons.html#search)?

Specifically, having a `exclude_addon_ids` or `exclude_addon_slugs` query parameter would be handy. These fields should allow comma-separated values to exclude multiple add-ons at once. Feel free to come up with a better name.

Thanks!

### LINKED ISSUE #6396: Consider indexing add-on slugs as 'keyword' in ES
At the moment, we index add-on slugs in ES as `text`, so they are analyzed. When searching, we include `slug` in the clauses.

But... does it make sense? For most add-ons, `slug` is just like a poor variant of `name`. Do we really need to analyze slugs and include them when searching with a query ? If slugs were indexed as keyword, it would allow exact matches on them, which seem more useful (allows to find an add-on by slug in ES, and also exclude them in #6376)

### PR DESCRIPTION
**Pull request by diox**
Fix #6376 
Fix #6396

### CODE REVIEW

File: docs/topics/api/addons.rst [hunk 1]
```diff
@@ -44,6 +44,7 @@ This endpoint allows you to search through public add-ons.
     :query string appversion: Filter by application version compatibility. Pass the full version as a string, e.g. ``46.0``. Only valid when the ``app`` parameter is also present.
     :query string author: Filter by exact author username. Multiple author names can be specified, separated by comma(s), in which case add-ons with at least one matching author are returned.
     :query string category: Filter by :ref:`category slug <category-list>`. ``app`` and ``type`` parameters need to be set, otherwise this parameter is ignored.
+    :query string exclude_addons: Exclude add-ons by ``slug`` or ``id``. Multiple add-ons can be specified, separated by comma(s).
```
Comment (EnTeQuAk) [inline docs/topics/api/addons.rst]: Stupid question just by reading the docs: Can `slug` and `id` be mixed? 😀 (which I know is explicitly tested in the unittest below)

File: src/olympia/addons/indexers.py [hunk 2]
```diff
@@ -159,7 +159,7 @@ def get_mapping(cls):
                             'average': {'type': 'float', 'index': False}
                         }
                     },
-                    'slug': {'type': 'text'},
+                    'slug': {'type': 'keyword'},
```
Comment (EnTeQuAk) [inline src/olympia/addons/indexers.py]: This needs a reindex unfortunately

File: src/olympia/search/filters.py [hunk 3]
```diff
@@ -67,7 +67,7 @@ def get_value_from_reverse_dict(self):
         return self.get_value_from_object_from_reverse_dict()
 
 
-class AddonAppVersionFilterParam(AddonFilterParam):
+class AddonAppVersionFilterParam(AddonQueryParam):
```
Comment (EnTeQuAk) [inline src/olympia/search/filters.py]: mhmm, naming… if it's inheriting from `AddonQueryParam` it should be a `AddonAppVersionQueryParam` imho. I know, that just blows up the diff, I'm fine doing that later.

File: src/olympia/search/filters.py [hunk 4]
```diff
@@ -175,20 +175,38 @@ def get_value_from_reverse_dict(self):
...
         return self.request.GET.get(self.query_param, '').split(',')
 
-    def get_es_filter(self):
+    def get_es_query(self):
         # Just using 'terms' would not work, as it would return any tag match
         # in the list, but we want to exactly match all of them.
         return [Q('term', tags=tag) for tag in self.get_value()]
 
 
+class AddonExcludeAddonsExcludeParam(AddonQueryParam):
```
Comment (EnTeQuAk) [inline src/olympia/search/filters.py]: Either continue naming this `FilterParam` or `QueryParam` but not just `Param` :)

File: src/olympia/search/filters.py [hunk 5]
```diff
@@ -175,20 +175,38 @@ def get_value_from_reverse_dict(self):
...
+    query_param = 'exclude_addons'
+
+    def get_value(self):
+        return self.request.GET.get(self.query_param, '').split(',')
```
Comment (EnTeQuAk) [inline src/olympia/search/filters.py]: we never trim the values for whitespaces, I wonder if that's something we should document or not.

File: src/olympia/search/filters.py [hunk 6]
```diff
@@ -327,29 +345,44 @@ def filter_queryset(self, request, qs, view):
...
+                         AddonPlatformFilterParam, AddonTagFilterParam,
+                         AddonTypeFilterParam]
 
-    def filter_queryset(self, request, qs, view):
-        must = []
+    available_excludes = [AddonExcludeAddonsExcludeParam]
 
-        for filter_class in self.available_filters:
+    def get_applicable_clauses(self, request, params_to_try):
+        rval = []
```
Comment (EnTeQuAk) [inline src/olympia/search/filters.py]: I'm a fan of naming this `return_value` or at least `retval` if it doesn't fit easily into 79 limit.

File: docs/topics/api/addons.rst (hunk 1 above)
Comment (diox) [reply docs/topics/api/addons.rst]: Do you think we need to word that differently to make it clear it's possible? IMHO it's not really worth it, it's supported but we don't really want to encourage it.

File: src/olympia/addons/indexers.py (hunk 2 above)
Comment (diox) [reply src/olympia/addons/indexers.py]: Yes, but that's fine, it's somewhat backwards-compatible, in the sense that no existing functionality breaks - it's just that slug-based exclusions will not work until the reindex.

File: docs/topics/api/addons.rst (hunk 1 above)
Comment (EnTeQuAk) [reply docs/topics/api/addons.rst]: hmm, we definitely don't want to encourage that. Maybe ignore me and let it the way it is.

**Pull request merged**